File: src/main/scala/com/ft/notificationsmonitor/PushReader.scala
package com.ft.notificationsmonitor

import java.time.ZonedDateTime

import akka.Done
import akka.actor.{Actor, ActorLogging, ActorRef, PoisonPill, Props}
import akka.stream.scaladsl.{Keep, Sink, Source}
import akka.stream.{ActorMaterializer, KillSwitches}
import akka.util.ByteString
import com.ft.notificationsmonitor.model.{DatedEntry, PushEntry}
import com.ft.notificationsmonitor.model.NotificationFormats._
import spray.json.DefaultJsonProtocol._
import spray.json._

import scala.concurrent.{Future, Promise}
import scala.util.{Failure, Success}

class PushReader(private val pairMatcher: ActorRef) extends Actor with ActorLogging {

  implicit private val ec = context.dispatcher
  implicit private val mat = ActorMaterializer()

  private val willCancelStreamP = Promise[Done]()

  override def receive: Receive = {
    case source: Source =>
      val (killSwitch, doneF) = consumeBodyStream(source)

      willCancelStreamP.future.onComplete { _ =>
        killSwitch.shutdown()
        self ! PoisonPill
      }

      doneF.onComplete { _ =>
        log.info("Stream has ended.")
        context.parent ! "StreamEnded"
      }

    case "CancelStreams" =>
      willCancelStreamP.complete(Success(Done))
  }

  private def consumeBodyStream(body: Source[ByteString, Any]) = {
    body.viaMat(KillSwitches.single)(Keep.right)
      .fold(ByteString(""))(foldPerLine)
      .toMat(Sink.ignore)(Keep.both)
      .run()
  }

  private def foldPerLine(acc: ByteString, next: ByteString) = {
    val nextString = next.utf8String
    nextString.lastIndexOf("\n\n") match {
      case -1 =>
        System.out.print(nextString + "|")
        acc ++ next
      case i =>
        val lines = nextString.split("\n\n").toList
        if (nextString.length - 2 == i) {
          parseLines(lines)
          ByteString("")
        } else {
          parseLines(lines.dropRight(1))
          ByteString(lines.last)
        }
    }
  }

  private def parseLines(lines: List[String]) = {
    lines.map(_.stripPrefix("data: [").stripSuffix("]")).foreach {
      case "" => log.info("heartbeat")
      case s => parseLine(s)
    }
  }

  private def parseLine(line: String) = {
    Future {
      val jsonLine = line.parseJson
      jsonLine.convertTo[PushEntry]
    }.onComplete {
      case Success(entry) =>
        log.info(entry.id)
        pairMatcher ! DatedEntry(entry, ZonedDateTime.now)
      case Failure(t) => log.error(t, "Error deserializing notifications response")
    }
  }
}

object PushReader {

  def props(pairMatcher: ActorRef) = Props(new PushReader(pairMatcher))

  case class Read(body: Source[ByteString, Any])
}

File: src/main/java/com/ft/notificationsmonitor/PushReader.java
package com.ft.notificationsmonitor;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Status.Failure;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Creator;
import akka.util.ByteString;
import com.ft.notificationsmonitor.model.DatedEntry;
import com.ft.notificationsmonitor.model.NotificationFormats;
import spray.json.JsValue;
import spray.json.JsonParser;
import spray.json.ParserInput;

import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static com.ft.notificationsmonitor.PushConnector.READER_FAILED;
import static com.ft.notificationsmonitor.PushConnector.RECONNECT;
import static java.time.format.DateTimeFormatter.ISO_INSTANT;

public class PushReader extends UntypedActor {

    static final String INIT = "Init";
    static final String ACK = "Ack";
    static final String COMPLETE = "Complete";
    static final String CHECK_HEARTBEAT = "CheckHeartbeat";
    static final Pattern INNER_JSON_ENTRY = Pattern.compile(".*data: \\[(.*)\\].*");

    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    private ActorRef pairMatcher;
    private ZonedDateTime heartbeat = ZonedDateTime.now();

    public PushReader(ActorRef pairMatcher) {
        this.pairMatcher = pairMatcher;
    }

    @Override
    public void onReceive(Object message) throws Throwable {
        if (message instanceof ByteString) {
            parseBytes((ByteString) message);
            sender().tell(ACK, self());

        } else if (message.equals(CHECK_HEARTBEAT)) {
            if (heartbeat.isBefore(ZonedDateTime.now().minusMinutes(1))) {
                log.warning("Missed out heartbeats for more than 1 minutes.");
                context().parent().tell(RECONNECT, self());
            }

        } else if (message.equals(INIT)) {
            getSender().tell(ACK, self());

        } else if (message.equals(COMPLETE)) {
            log.info("Stream has ended.");
            context().parent().tell(RECONNECT, self());

        } else if (message instanceof Failure) {
            final Failure failure = (Failure) message;
            log.error(failure.cause(), "Stream has failed.");
            getContext().parent().tell(READER_FAILED, getSelf());
        }
    }

    private void parseBytes(final ByteString byteString) {
        final String line = byteString.utf8String();
        final String unwrappedLine;
        Matcher innerJsonMatcher = INNER_JSON_ENTRY.matcher(line);
        if (innerJsonMatcher.find()) {
            unwrappedLine = innerJsonMatcher.group(1);
        } else {
            log.warning("not matching 'data: []' pattern line={}", line);
            return;
        }
        heartbeat = ZonedDateTime.now();
        if (unwrappedLine.isEmpty()) {
            log.info("heartbeat");
        } else {
            parseUnwrappedLine(unwrappedLine);
        }
    }

    private void parseUnwrappedLine(final String line) {
        System.out.println(line);
        CompletableFuture.supplyAsync(() -> {
            JsValue parser = JsonParser.apply(new ParserInput.StringBasedParserInput(line));
            return parser.convertTo(NotificationFormats.pushEntryFormat());
        }).whenComplete((entry, failure) -> {
            if (failure != null) {
                log.error(failure, "Error deserializing notifications response");
            } else {
                final DatedEntry datedEntry = new DatedEntry(entry, ZonedDateTime.now());
                log.info("id={} publishReference={} lastModified=\"{}\" foundAt=\"{}\"",
                        entry.id(),
                        entry.publishReference(),
                        entry.lastModified().format(ISO_INSTANT),
                        datedEntry.getDate().format(ISO_INSTANT));
                pairMatcher.tell(datedEntry, self());
            }
        });
    }

    public static Props props(final ActorRef pairMatcher) {
        return Props.create(new Creator<PushReader>() {
            private static final long serialVersionUID = 1L;

            @Override
            public PushReader create() throws Exception {
                return new PushReader(pairMatcher);
            }
        });
    }
}
